scanner/src/main/kotlin/experimental/NestedProvenanceScanResult.kt

```diff
@@ -136,4 +137,75 @@ data class NestedProvenanceScanResult(
 
         throw IllegalArgumentException("Could not find entry for $provenance.")
     }
+
+    /**
+     * Remove all scan results for [nestedProvenance]s which are not within the provided [path] and filter all findings
+     * within the scan results by [path], taking the sub repository paths into account.
+     * This also updates the VCS paths for all [RepositoryProvenance]s accordingly to make clear that they are not
+     * results for the whole repository.
+     *
+     * Throws an [IllegalArgumentException] if any provenance in [nestedProvenance] already has a VCS path set.
+     */
+    fun filterByVcsPath(path: String): NestedProvenanceScanResult {
+        if (path.isBlank()) return this
+
+        val provenances = getProvenances()
+        val provenancesWithVcsPath = provenances.filter { it is RepositoryProvenance && it.vcsInfo.path.isNotBlank() }
+
+        require(provenancesWithVcsPath.isEmpty()) {
+            "Cannot filter scan results by VCS path that have a repository provenance with a non-blank VCS path, " +
+                    "because the partial scan result might not contain the path to filter by, but the following " +
+                    "provenances have a non-blank VCS path: ${provenancesWithVcsPath.joinToString("\n") { "\t$it" }}."
+        }
+
+        val pathsWithinProvenances = provenances.filter {
+            val provenancePath = getPath(it)
+            provenancePath.isBlank() || path.startsWith("$provenancePath/")
+        }.associateWith { provenance ->
+            val provenancePath = getPath(provenance)
+            path.removePrefix("$provenancePath/")
+        }
+
+        fun KnownProvenance.withVcsPath() =
+            when (this) {
+                is RepositoryProvenance -> {
+                    val pathWithinProvenance = pathsWithinProvenances.getValue(this)
+                    copy(vcsInfo = vcsInfo.copy(path = pathWithinProvenance))
+                }
+
+                else -> this
+            }
+
+        // Set the VCS path of all provenances according to the provided path and filter provenances which are not
+        // within path.
+        val newNestedProvenance = NestedProvenance(
+            root = nestedProvenance.root.withVcsPath(),
+            subRepositories = nestedProvenance.subRepositories.filterValues { it in pathsWithinProvenances.keys }
+                .mapValues { (_, provenance) -> provenance.withVcsPath() as RepositoryProvenance }
+        )
```

Comment:
Can you help me understanding? ...why it is necessary to set the VCS path in the returned NestedProvenanceScanResult?

If the VCS path of the root provenance points into a subdirectory of a submodule, this is the case when it is done IIUC. If the goal is to scan the full repo no matter of the VCS path, why is setting this path needed?

Reply:
Thanks!